javascript/src/html_processor.ts

```diff
@@ -163,21 +168,82 @@ function actionForElement(element: Element): DomAction {
     switch (style.whiteSpace) {
       case 'nowrap':
       case 'pre':
-        return DomAction.Skip;
+        return DomAction.NoBreak;
     }
 
     const display = style.display;
     if (display)
       return display === 'inline' ? DomAction.Inline : DomAction.Block;
     // `display` is an empty string if the element is not connected.
   }
+
   // Use the built-in rules if the `display` property is empty, or if
   // `getComputedStyle` is missing (e.g., jsdom.)
   return defaultBlockElements.has(nodeName)
     ? DomAction.Block
     : DomAction.Inline;
 }
 
+/**
+ * Represents a node in {@link Paragraph}.
+ *
+ * It wraps a {@link Text} or a {@link string}.
+ *
+ * A {@link string} provides the context for the parser, but it can't be split.
+ */
+export class NodeOrText {
+  nodeOrText: Text | string;
+  chunks: string[] = [];
+
+  constructor(nodeOrText: Text | string) {
+    this.nodeOrText = nodeOrText;
+  }
+
+  get isString(): boolean {
+    return typeof this.nodeOrText === 'string';
+  }
+  get canSplit(): boolean {
+    return !this.isString;
+  }
+  get text(): string | null {
+    return this.isString
+      ? (this.nodeOrText as string)
+      : (this.nodeOrText as Text).nodeValue;
+  }
+  get length(): number {
+    return this.text?.length ?? 0;
+  }
+
+  /**
+   * Split the {@link Text} in the same way as the {@link chunks}.
+   * Joining all {@link chunks} must be equal to {@link text}.
+   */
+  split(separator: string | Node) {
+    const chunks = this.chunks;
+    assert(chunks.length === 0 || chunks.join('') === this.text);
+    if (chunks.length <= 1) return;
+    assert(this.canSplit);
+    const node = this.nodeOrText as Text;
+    if (typeof separator === 'string') {
+      // If the `separator` is a string, insert it at each boundary.
+      node.nodeValue = chunks.join(separator);
+      return;
+    }
+
+    // Otherwise create a `Text` node for each chunk, with the separator node
+    // between them, and replace the `node` with them.
+    const document = node.ownerDocument;
+    let nodes = [];
+    for (const chunk of chunks) {
+      if (chunk) nodes.push(document.createTextNode(chunk));
+      nodes.push(null);
```

Comment:
Is there any specific reason to not insert a cloned separator element here directly?

Reply:
It's to avoid unnecessary `cloneNode`. Added comments.

Reply:
Makes sense! Thank you for adding the comment.